lib/Interpreter/CIFactory.cpp

```diff
@@ -999,35 +1140,46 @@ namespace {
       }
 #endif
 
-    argvCompile.push_back("-c");
-    argvCompile.push_back("-");
+    const bool genOutput = COpts.HasOutput && !OnlyLex;
+    if (!genOutput) {
+      argvCompile.push_back("-c");
+      argvCompile.push_back("-");
+    }
+
+    std::unique_ptr<clang::CompilerInvocation>
+      InvocationPtr(new clang::CompilerInvocation);
+    if (!InvocationPtr)
+      return nullptr;
 
-    clang::CompilerInvocation*
-      Invocation = new clang::CompilerInvocation;
     // The compiler invocation is the owner of the diagnostic options.
     // Everything else points to them.
-    DiagnosticOptions& DiagOpts = Invocation->getDiagnosticOpts();
-    TextDiagnosticPrinter* DiagnosticPrinter
-      = new TextDiagnosticPrinter(llvm::errs(), &DiagOpts);
-    llvm::IntrusiveRefCntPtr<clang::DiagnosticIDs> DiagIDs(new DiagnosticIDs());
-    llvm::IntrusiveRefCntPtr<DiagnosticsEngine>
-      Diags(new DiagnosticsEngine(DiagIDs, &DiagOpts,
-                                  DiagnosticPrinter, /*Owns it*/ true));
+    DiagnosticOptions& DiagOpts = InvocationPtr->getDiagnosticOpts();
+    llvm::IntrusiveRefCntPtr<DiagnosticsEngine> Diags =
+        SetupDiagnostics(DiagOpts);
+    if (!Diags) {
+      // If we can't even setup the diagnostic engine, lets not use llvm::errs
+      ::perror("Could not setup diagnostic engine");
+      return nullptr;
+    }
+
     clang::driver::Driver Driver(argv[0], llvm::sys::getDefaultTargetTriple(),
                                  *Diags);
     //Driver.setWarnMissingInput(false);
     Driver.setCheckInputsExist(false); // think foo.C(12)
     llvm::ArrayRef<const char*>RF(&(argvCompile[0]), argvCompile.size());
     std::unique_ptr<clang::driver::Compilation>
       Compilation(Driver.BuildCompilation(RF));
-    const clang::driver::ArgStringList* CC1Args
-      = GetCC1Arguments(Diags.get(), Compilation.get());
-    if (CC1Args == NULL) {
-      delete Invocation;
-      return 0;
+    if (!Compilation) {
```

Comment:
Can we use a variable name that is 'distinct' from a class name (to avoid confusion between constructor and variable)?